### DISCUSSION TITLE
Can it code?

### DISCUSSION THREAD

**Discussion by ARandomUserFromGithub** [oobabooga/text-generation-webui#183] [Q&A]
I tried this simple experiment... It display large blue text, which is impressive, but it only accepted to do it once.
`     User:  "Can you give me a CSS example which would make H1 text bold, blue and 12em",
  Chiharu Yamada:    "Sure thing! *Chiharu writes some code and then shows you the result* Here you go! <h1 style=\"font-weight:bold;color:#0000ff;\">Example</h1>"`
  
Is security strong enough for the AI to write code out of the wild outside of proper code embeds?

I'm not super experienced with huggingface but what models should I download for more advanced tools that could teach me code?
I'm also looking forward to get a good set of tools to translate. ChatGPT can translate flawlessly, and I would like a model that can do just that.

I really appreciate TGwebUI, it's awesome!

Looking forward to the future of it!

New discovery: It can also display images, whereas it displayed Google's logo while trying to display am emoji, which failed... Which made me discover it can display images... Sounds interesting. ... On a new try, it worked. Displayed a nice emoji... But couldn't display the photo of the current president of Congo.

The prompt "Can you list several emojis that are popular in your training data?" is really interesting.. It makes the AI draw a table with actual html code.  Could that awesome feature where it writes HTML be foolproofed so it won't pose security risk ever?

The prompt: `Can you display code inside of a code embedding?` made it put actual code iside of an embedding! Maybe I could ask it to do it in the character context or something... IT's so awesome!

Now the only reply it knows anymore is: "Sure thing! Here we go..." lmao... It made 1 embedding 1 time, then after trying to turn it into a character... All it does anymore is say "here we go" and nothing more.

Will there be a library of user-made presets and characters? It would be so great!

**Comment by oobabooga** [comment]
> Is security strong enough for the AI to write code out of the wild outside of proper code embeds?

Honestly, if you manage to get pwned by HTML code outputted by the AI into the chat box, please let me know. That would be very amusing.

**Reply by ARandomUserFromGithub** [reply]
I managed to input my own web image links with `<img src="https://media.sproutsocial.com/uploads/2017/02/10x-featured-social-media-image-size.png">` and it managed to display an image by itself too, once. 
Maybe an addon could be developed for that purpose.

**Comment by ETA-22** [comment]
Yeah, also curious what models and generator presets work well for code. I’ve found ChatGPT helpful at times in discovery and debugging, but would prefer to have something local.

**Comment by ZeroIQ1024** [comment]
I downloaded the model "opt-6.7b" and asked it for simple hello world javascript code, just to see if it could understand what I was asking for and I swear to god its reply was "are you on acid?" LOL.

If anybody finds a good model for coding or what settings to use, please let me/us know :-)

**Reply by sinanisler** [reply]
some models pretty crazy :)
training a model on 4chan and reddit not a good idea :D

**Reply by klink** [reply]
you can try Refact LLM

**Reply by ye7iaserag** [reply]
Phind v2 has the highest HumanEval right now (73.8)

**Reply by ye7iaserag** [reply]
DeepSeek is the SOTA now

**Reply by TomLucidor** [reply]
Wait how do you run this vs CodeGeeX2, CodeGen, or CodeT5+?

**Comment by JammersUK** [comment]
I found a model the webgui can run: [codegen-2B-mono at ](https://huggingface.co/Salesforce/codegen-2B-mono)
I haven't worked out yet, how to get it to work properly, but it's a start.  The codet5 models just crashed on start up.

**Reply by sinanisler** [reply]
hmmm interesting...

**Reply by nerdyrodent** [reply]
Codegen seemed pretty good to me!

**Reply by JammersUK** [reply]
This isn't the same as GPT-4/Bing.  I just asked it to make a calculator in Python and it produced a fully working calculator, with no modification needed.
What we have are older/smaller models that require specific prompts to produce the code and then it might not write a full program in one go.  But for software running on a gaming graphics card on a small tower PC, it's doing a good job.

**Reply by ymonster** [reply]
I am stucking on how to run codegen model. When I run star-webui.bat it told me  "_**Can't determine model**_". I fount a 6b-gptj model on hf and set --model_type=gptj, Then it told me "**_can't find .pt or .safetensor file_**".

**Reply by TomLucidor** [reply]
Try CodeGeeX2 first since they have the same folder structure, not sure if it is good tho

**Comment by olihough86** [comment]
The Alpaca LoRA can write C, got a working hello, world out of it, not sure about complex code

**Comment by oobabooga** [comment]
It can write pretty complex Python code too
![img](https://user-images.githubusercontent.com/112222186/226108211-445219ec-1f8e-440f-b9b6-ebe49f54ed7f.png)

**Reply by FartyPants** [reply]
The bonus of llama is that it also knows bit more gradio than chatgpt as their cutoff time was earlier.
chatgpt basically makes stuff up in that regard.

**Reply by TomLucidor** [reply]
What are the settings needed for this vs using a code-specific LLM?

**Reply by JammersUK** [reply]
I have been using phind-codellama-34b-v2.Q4_K_M.gguf, with a lot of success.  

```
import subprocess
import pandas as pd
from datetime import datetime

# Name of CSV file - Check if it is in same directory as this program, if not change it
csv_file = ('enviro_log.csv')

# Read CSV file into a DataFrame
df = pd.read_csv(csv_file)

# Convert 'Timestamp' column to datetime type if it isn't already
df['Timestamp (datetime)'] = pd.to_datetime(df['Timestamp (datetime)'], dayfirst=True)

# Extract the date from 'Timestamp' column
df['Date'] = df['Timestamp (datetime)'].dt.date

# Group by date and calculate average humidity for each day
average_humidity = df.groupby('Date')['Humidity (%)'].mean()

# Display the result
for index, value in zip(average_humidity.index, average_humidity.values):
    print(f"Average humidity for the day {index} is {value:.2f}%")

# Replace 'script2.py' with the name of the script you want to run
result = subprocess.run(["python", "gas_detector.py"], capture_output=True, text=True)

print("The output from the gas detection script is:")
print(result.stdout)  # This will print the standard output from the second script
# End
```


The output:
Average humidity for the day 2023-11-30 is 31.03%
Average humidity for the day 2023-12-01 is 29.88%
Average humidity for the day 2023-12-02 is 24.14%

The output from the gas detection script is:
Date: 2023-11-30 10:30:10, Detected Gas: Nh3 (kO)
Date: 2023-11-30 10:30:10, Detected Gas: Nh3 (kO)

The only thing I changed was the script2.py to gas_detector.py.  Everything else is 100% AI.  So to me (I'm not a programmer), we can get good working Python code.

**Comment by TomLucidor** [comment]
These four models popped up and probs a good idea to test them before, there are VSCode extensions that these work on
- CodeGeeX2 https://github.com/THUDM/CodeGeeX2
- Starcoder https://github.com/bigcode-project/starcoder
- CodeGen https://github.com/salesforce/CodeGen
- CodeT5+ https://github.com/salesforce/CodeT5
